[NEXT-832] Dev Server NPM error code ENOWORKSPACES

### Verify canary release

- [X] I verified that the issue exists in the latest Next.js canary release

### Provide environment information

```bash
Operating System:
      Platform: darwin
      Arch: arm64
      Version: Darwin Kernel Version 22.2.0: Fri Nov 11 02:03:51 PST 2022; root:xnu-8792.61.2~4/RELEASE_ARM64_T6000
    Binaries:
      Node: 18.15.0
      npm: 9.6.1
      Yarn: N/A
      pnpm: N/A
    Relevant packages:
      next: 13.2.5-canary.3
      eslint-config-next: 13.2.4
      react: 18.2.0
      react-dom: 18.2.0
```


### Which area(s) of Next.js are affected? (leave empty if unsure)

Dev server and NPM

### Link to the code that reproduces this issue

https://github.com/joepetrillo/workspace-issue

### To Reproduce

Clone the repo then `npm install` then `cd reproduction-app` then `npm run dev`

<img width="701" alt="image" src="https://user-images.githubusercontent.com/32227357/225073547-009f5219-3aa0-478c-afaf-0ae1bef23e77.png">

### Describe the Bug

**I have determined that this error happens when a nextjs (versions after 13.1.6) dev server runs in an npm workspace.**

`npm config get registry` is getting called sometime during the dev server startup and is throwing an error since it is not supported in workspaces. 

`error code ENOWORKSPACES`
`error This command does not support workspaces.`

I initially thought this could be a turborepo issue, but it is not. More details are in the issue I initially opened (linked below).

https://github.com/vercel/turbo/issues/4183

And here is a sample error log:

```
0 verbose cli /Users/jpetrillo/Library/Application Support/fnm/node-versions/v18.15.0/installation/bin/node /Users/jpetrillo/Library/Caches/fnm_multishells/18620_1678811038524/bin/npm
1 info using npm@9.6.1
2 info using node@v18.15.0
3 timing npm:load:whichnode Completed in 0ms
4 timing config:load:defaults Completed in 0ms
5 timing config:load:file:/Users/jpetrillo/Library/Application Support/fnm/node-versions/v18.15.0/installation/lib/node_modules/npm/npmrc Completed in 1ms
6 timing config:load:builtin Completed in 1ms
7 timing config:load:cli Completed in 0ms
8 timing config:load:env Completed in 0ms
9 info found workspace root at /Users/jpetrillo/Documents/Projects/workspace-bug
10 timing config:load:file:/Users/jpetrillo/Documents/Projects/workspace-bug/.npmrc Completed in 0ms
11 timing config:load:project Completed in 3ms
12 timing config:load:file:/Users/jpetrillo/.npmrc Completed in 0ms
13 timing config:load:user Completed in 0ms
14 timing config:load:file:/Users/jpetrillo/Library/Application Support/fnm/node-versions/v18.15.0/installation/etc/npmrc Completed in 0ms
15 timing config:load:global Completed in 0ms
16 timing config:load:setEnvs Completed in 0ms
17 timing config:load Completed in 5ms
18 timing npm:load:configload Completed in 5ms
19 timing npm:load:mkdirpcache Completed in 0ms
20 timing npm:load:mkdirplogs Completed in 0ms
21 verbose title npm config get registry
22 verbose argv "config" "get" "registry"
23 timing npm:load:setTitle Completed in 5ms
24 timing config:load:flatten Completed in 1ms
25 timing npm:load:display Completed in 1ms
26 verbose logfile logs-max:10 dir:/Users/jpetrillo/.npm/_logs/2023-03-14T16_37_32_128Z-
27 verbose logfile /Users/jpetrillo/.npm/_logs/2023-03-14T16_37_32_128Z-debug-0.log
28 timing npm:load:logFile Completed in 2ms
29 timing npm:load:timers Completed in 0ms
30 timing npm:load:configScope Completed in 0ms
31 timing npm:load Completed in 15ms
32 timing command:config Completed in 0ms
33 verbose stack Error: This command does not support workspaces.
33 verbose stack     at Config.cmdExec (/Users/jpetrillo/Library/Application Support/fnm/node-versions/v18.15.0/installation/lib/node_modules/npm/lib/base-command.js:123:29)
33 verbose stack     at Npm.exec (/Users/jpetrillo/Library/Application Support/fnm/node-versions/v18.15.0/installation/lib/node_modules/npm/lib/npm.js:154:20)
33 verbose stack     at async module.exports (/Users/jpetrillo/Library/Application Support/fnm/node-versions/v18.15.0/installation/lib/node_modules/npm/lib/cli.js:134:5)
34 verbose cwd /Users/jpetrillo/Documents/Projects/workspace-bug/reproduction-app
35 verbose Darwin 22.2.0
36 verbose node v18.15.0
37 verbose npm  v9.6.1
38 error code ENOWORKSPACES
39 error This command does not support workspaces.
40 verbose exit 1
41 timing npm Completed in 21ms
42 verbose code 1
43 error A complete log of this run can be found in:
43 error     /Users/jpetrillo/.npm/_logs/2023-03-14T16_37_32_128Z-debug-0.log
```

### Expected Behavior

No error

### Which browser are you using? (if relevant)

_No response_

### How are you deploying your application? (if relevant)

_No response_

<sub>[NEXT-832](https://linear.app/vercel/issue/NEXT-832/dev-server-npm-error-code-enoworkspaces)</sub>

I get the error with 3.1.6 also on first dev server start, seemingly it tries to patch lock file

```
ready - started server on 0.0.0.0:3030, url: http://localhost:3030
info  - Loaded env from /Users/jibyjose/Projects/aurum-one/web/dashboard/.env
warn  - The `app` directory is experimental. To enable, add `appDir: true` to your `next.config.js` configuration under `experimental`. See https://nextjs.org/docs/messages/experimental-app-dir-config
warn  - The `app` directory is experimental. To enable, add `appDir: true` to your `next.config.js` configuration under `experimental`. See https://nextjs.org/docs/messages/experimental-app-dir-config
info  - automatically enabled Fast Refresh for 1 custom loader
warn  - Found lockfile missing swc dependencies, patching...
npm ERR! code ENOWORKSPACES
npm ERR! This command does not support workspaces.

npm ERR! A complete log of this run can be found in:
npm ERR!     /Users/jibyjose/.npm/_logs/2023-03-28T13_55_04_964Z-debug-0.log
event - compiled client and server successfully in 2s (4135 modules)
warn  - Lockfile was successfully patched, please run "npm install" to ensure @next/swc dependencies are downloaded
```

I also receive the same error when I spin up a dev server on a turborepo/next/npm project. Everything still builds and runs like @joepetrillo stated.

Error:
```shell
npm ERR! code ENOWORKSPACES
npm ERR! This command does not support workspaces.

npm ERR! A complete log of this run can be found in: ...
```

Package versions:
```json
"next": "13.2.4",
"react": "18.2.0",
"react-dom": "18.2.0",
"turbo": "1.8.3",
```

node/npm versions:
```shell
node --version
v18.15.0

npm --version
9.5.0
```

--- a/packages/next/src/lib/helpers/get-registry.ts
+++ b/packages/next/src/lib/helpers/get-registry.ts
@@ -23,7 +23,11 @@ export function getRegistry(baseDir: string = process.cwd()) {
     if (output.startsWith('http')) {
       registry = output.endsWith('/') ? output : `${output}/`
     }
-  } finally {
-    return registry
+  } catch {
+    // ignore error
+    // e.g. In an npm workspace, `npm config get registry` will throw error code ENOWORKSPACES
+    // x-ref: https://github.com/vercel/next.js/issues/47121#issuecomment-1499044345
   }
+
+  return registry
 }

--- a/packages/next/src/lib/helpers/get-registry.ts
+++ b/packages/next/src/lib/helpers/get-registry.ts
@@ -35,9 +35,9 @@ export function getRegistry(baseDir: string = process.cwd()) {
     }
   } catch (err) {
     if (err instanceof Error) {
-      throw new Error(
-        `Failed to get registry from "${pkgManager}".\n\n${err.message}`
-      )
+      throw new Error(`Failed to get registry from "${pkgManager}".`, {
+        cause: err,
+      })
     }
   }
 

--- a/packages/next/src/lib/helpers/get-registry.ts
+++ b/packages/next/src/lib/helpers/get-registry.ts
@@ -42,11 +42,9 @@ export function getRegistry(baseDir: string = process.cwd()) {
     }
 
     try {
-      // x-ref: https://github.com/npm/cli/issues/6099#issuecomment-1847584792
-      const output = runConfigGetRegistry(pkgManager, [
-        '--workspaces=false',
-        '--include-workspace-root',
-      ])
+      // run command under the context of the root project only
+      // x-ref: https://github.com/npm/statusboard/issues/371#issue-920669998
+      const output = runConfigGetRegistry(pkgManager, ['--no-workspaces'])
 
       if (output.startsWith('http')) {
         registry = addTrailingSlash(output)

--- a/packages/next/src/lib/helpers/get-registry.ts
+++ b/packages/next/src/lib/helpers/get-registry.ts
@@ -34,11 +34,9 @@ export function getRegistry(baseDir: string = process.cwd()) {
       registry = output.endsWith('/') ? output : `${output}/`
     }
   } catch (err) {
-    if (err instanceof Error) {
-      throw new Error(`Failed to get registry from "${pkgManager}".`, {
-        cause: err,
-      })
-    }
+    throw new Error(`Failed to get registry from "${pkgManager}".`, {
+      cause: err,
+    })
   }
 
   return registry

--- a/packages/next/src/lib/helpers/get-registry.ts
+++ b/packages/next/src/lib/helpers/get-registry.ts
@@ -1,23 +1,6 @@
-import isError from '../is-error'
 import { execSync } from 'child_process'
 import { getPkgManager } from './get-pkg-manager'
 import { getFormattedNodeOptionsWithoutInspect } from '../../server/lib/utils'
-
-function runConfigGetRegistry(pkgManager: string, flags?: string[]) {
-  const resolvedFlags = flags ? flags.join(' ') : ''
-  return execSync(`${pkgManager} config get registry ${resolvedFlags}`, {
-    env: {
-      ...process.env,
-      NODE_OPTIONS: getFormattedNodeOptionsWithoutInspect(),
-    },
-  })
-    .toString()
-    .trim()
-}
-
-function addTrailingSlash(url: string) {
-  return url.endsWith('/') ? url : `${url}/`
-}
 
 /**
  * Returns the package registry using the user's package manager.
@@ -26,33 +9,31 @@ function addTrailingSlash(url: string) {
  */
 export function getRegistry(baseDir: string = process.cwd()) {
   const pkgManager = getPkgManager(baseDir)
+  // Since `npm config` command fails in npm workspace to prevent workspace config conflicts,
+  // add `--no-workspaces` flag to run under the context of the root project only.
+  // Safe for non-workspace projects as it's equivalent to default `--workspaces=false`.
+  // x-ref: https://github.com/vercel/next.js/issues/47121#issuecomment-1499044345
+  // x-ref: https://github.com/npm/statusboard/issues/371#issue-920669998
+  const resolvedFlags = pkgManager === 'npm' ? '--no-workspaces' : ''
   let registry = `https://registry.npmjs.org/`
 
   try {
-    const output = runConfigGetRegistry(pkgManager)
+    const output = execSync(
+      `${pkgManager} config get registry ${resolvedFlags}`,
+      {
+        env: {
+          ...process.env,
+          NODE_OPTIONS: getFormattedNodeOptionsWithoutInspect(),
+        },
+      }
+    )
+      .toString()
+      .trim()
 
     if (output.startsWith('http')) {
-      registry = addTrailingSlash(output)
+      registry = output.endsWith('/') ? output : `${output}/`
     }
-  } catch (error) {
-    // `npm config` command fails in npm workspace to prevent workspace config conflicts.
-    // x-ref: https://github.com/vercel/next.js/issues/47121#issuecomment-1499044345
-    if (isError(error) && error.code !== 'ENOWORKSPACES') {
-      throw error
-    }
-
-    try {
-      // run command under the context of the root project only
-      // x-ref: https://github.com/npm/statusboard/issues/371#issue-920669998
-      const output = runConfigGetRegistry(pkgManager, ['--no-workspaces'])
-
-      if (output.startsWith('http')) {
-        registry = addTrailingSlash(output)
-      }
-    } catch (e) {
-      throw e
-    }
+  } finally {
+    return registry
   }
-
-  return registry
 }

--- a/packages/next/src/lib/helpers/get-registry.ts
+++ b/packages/next/src/lib/helpers/get-registry.ts
@@ -3,33 +3,56 @@ import { execSync } from 'child_process'
 import { getPkgManager } from './get-pkg-manager'
 import { getFormattedNodeOptionsWithoutInspect } from '../../server/lib/utils'
 
+function runConfigGetRegistry(pkgManager: string, flags?: string[]) {
+  const resolvedFlags = flags ? flags.join(' ') : ''
+  return execSync(`${pkgManager} config get registry ${resolvedFlags}`, {
+    env: {
+      ...process.env,
+      NODE_OPTIONS: getFormattedNodeOptionsWithoutInspect(),
+    },
+  })
+    .toString()
+    .trim()
+}
+
+function addTrailingSlash(url: string) {
+  return url.endsWith('/') ? url : `${url}/`
+}
+
 /**
  * Returns the package registry using the user's package manager.
  * The URL will have a trailing slash.
  * @default https://registry.npmjs.org/
  */
 export function getRegistry(baseDir: string = process.cwd()) {
+  const pkgManager = getPkgManager(baseDir)
   let registry = `https://registry.npmjs.org/`
+
   try {
-    const pkgManager = getPkgManager(baseDir)
-    const output = execSync(`${pkgManager} config get registry`, {
-      env: {
-        ...process.env,
-        NODE_OPTIONS: getFormattedNodeOptionsWithoutInspect(),
-      },
-    })
-      .toString()
-      .trim()
+    const output = runConfigGetRegistry(pkgManager)
 
     if (output.startsWith('http')) {
-      registry = output.endsWith('/') ? output : `${output}/`
+      registry = addTrailingSlash(output)
     }
   } catch (error) {
-    // In an npm workspace, `npm config get registry` will throw error code ENOWORKSPACES
-    // As this is NPM specific error, we ignore and use the default NPM registry.
+    // `npm config` command fails in npm workspace to prevent workspace config conflicts.
     // x-ref: https://github.com/vercel/next.js/issues/47121#issuecomment-1499044345
     if (isError(error) && error.code !== 'ENOWORKSPACES') {
       throw error
+    }
+
+    try {
+      // x-ref: https://github.com/npm/cli/issues/6099#issuecomment-1847584792
+      const output = runConfigGetRegistry(pkgManager, [
+        '--workspaces=false',
+        '--include-workspace-root',
+      ])
+
+      if (output.startsWith('http')) {
+        registry = addTrailingSlash(output)
+      }
+    } catch (e) {
+      throw e
     }
   }
 

--- a/packages/next/src/lib/helpers/get-registry.ts
+++ b/packages/next/src/lib/helpers/get-registry.ts
@@ -1,3 +1,4 @@
+import isError from '../is-error'
 import { execSync } from 'child_process'
 import { getPkgManager } from './get-pkg-manager'
 import { getFormattedNodeOptionsWithoutInspect } from '../../server/lib/utils'
@@ -23,10 +24,13 @@ export function getRegistry(baseDir: string = process.cwd()) {
     if (output.startsWith('http')) {
       registry = output.endsWith('/') ? output : `${output}/`
     }
-  } catch {
-    // ignore error
-    // e.g. In an npm workspace, `npm config get registry` will throw error code ENOWORKSPACES
+  } catch (error) {
+    // In an npm workspace, `npm config get registry` will throw error code ENOWORKSPACES
+    // As this is NPM specific error, we ignore and use the default NPM registry.
     // x-ref: https://github.com/vercel/next.js/issues/47121#issuecomment-1499044345
+    if (isError(error) && error.code !== 'ENOWORKSPACES') {
+      throw error
+    }
   }
 
   return registry

--- a/packages/next/src/lib/helpers/get-registry.ts
+++ b/packages/next/src/lib/helpers/get-registry.ts
@@ -33,7 +33,13 @@ export function getRegistry(baseDir: string = process.cwd()) {
     if (output.startsWith('http')) {
       registry = output.endsWith('/') ? output : `${output}/`
     }
-  } finally {
-    return registry
+  } catch (err) {
+    if (err instanceof Error) {
+      throw new Error(
+        `Failed to get registry from "${pkgManager}".\n\n${err.message}`
+      )
+    }
   }
+
+  return registry
 }

